[Mobile Payments] Show spinner while awaiting reader readiness

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8079
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

The `Reader is ready` screen used to show immediately when you tap `Card` on the Payment Method screen, to start the card payment, when a card reader is connected, or immediately after a reader is connected.

This led to a slightly jarring payment experience for the merchant and their customers, as we prompted the merchant to ask their customer to tap their card before the reader is actually ready. There are network-bound operations which happen in this state, so the delay could be tens of seconds in a bad case. On fast Wifi, it's still noticeable at 1-2 seconds of delay

In this PR, we add a `Getting ready to collect payment` screen with a spinner at this point, and then show the `Reader is ready` screen when the Payment Intent has been returned. 

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Navigate to `Menu > Payments > Collect Payment`
Enter an amount, tap `Next`, then `Take Payment`.
Be ready to tap your card immediately on seeing `Reader is ready` on the phone
Select `Card` as the payment method.
Observe that a spinner shows for a few seconds.
Observe that `Reader is ready` then shows on the phone, at the same time as the green light goes steady on the M2 reader.
Observe that the card can be read immediately.

To see this more clearly, set a breakpoint in Proxyman/Charles on `https://api.stripe.com/v1/payment_intents`, and observe that `Reader is ready` displays before that returns, but you can't tap your card until after you let the breakpoint run.

N.B. This affects the Interac refund flow as well, so Interac payment and refunds should be tested. Try both with and without a connected reader.

I noticed that the `Tap/Swipe/Insert` card alert would not work in the current code, and logged #8116 to fix this. This is not a regression of this PR.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

https://user-images.githubusercontent.com/2472348/201758295-b77d63d4-db75-4e27-aad1-d45bc60741a6.mp4

N.B. ignore the tipping screen: it's leftover from the WisePad configuration I added in my HACK week project and won't show for anyone else.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

joshheald committed:
8079 Add Indefinite progress view style

joshheald committed:
8079Add PreparingReader card reader modal

joshheald committed:
8079 Show spinner while preparing reader for payment

joshheald committed:
Update tests to match changes to flow

joshheald committed:
8079 Add Preparing Card Reader spinner to 11.3

joshheald commented:
@joe-keenan Here's the fix with the spinner I mentioned. Let me know what you think...

joshheald commented:
@lmischner @jostnes Optional review on this one... it's preparation which will make the Tap on Mobile UI much smoother, so can be reviewed as part of the overall flow later if that's more convenient for you.

koke reviewed (approved):
This seems to work fine with the M2 reader. Note that I don't have a WisePad 3 anymore so I can't test Interac. Once thing I noticed is that the cancel button on the new modal will dismiss the modal, but then it will show the reader-is-ready modal again once ready. I'm not sure how much of a blocker it is but we might want to disable that cancel button if it doesn't actually cancel the reader initialization. 

![cancel-reader-ready](https://user-images.githubusercontent.com/8739/201854077-8e700212-d83c-4be1-9e18-b87be3e2f495.gif)

joshheald commented:
> This seems to work fine with the M2 reader. Note that I don't have a WisePad 3 anymore so I can't test Interac. 

Thanks for the review @koke 

> One thing I noticed is that the cancel button on the new modal will dismiss the modal, but then it will show the reader-is-ready modal again once ready. 

I saw this but only once, and I thought I'd fixed it with another change. :/ if you're still seeing it, it's a bigger issue than I thought. 

> I'm not sure how much of a blocker it is but we might want to disable that cancel button if it doesn't actually cancel the reader initialization.

I'll see if I can figure out what it is – I'd rather have the cancel button as a spinner with no way out other than force quitting the app seems bad. That said, there are other steps in the flow which can't be cancelled either, so it's not a lot worse as a solution.

joshheald committed:
Add feature area to preparing reader release note

joshheald committed:
8079 Remove plugins.sync to fix payment cancellation

The `ReceiptEmailParameterDeterminer` synchronized plugins every time `receiptEmail` was called, i.e. when setting up the payment parameters at the start of every payment flow. This added an async call to the reader preparation flow, and if the flow was cancelled during that flow, the payment continued.

This happened because `CardPresentPaymentAction.collectPayment` hadn’t been sent yet, so we didn’t have a payment to cancel. This was sent when the plugin sync completed, even if we’d cancelled the flow in the meantime.

We also sync these plugins in the onboarding flow, which is checked at the start of every payment. Syncing again on the payment flow’s critical path is unneccesary.

This change removes the extra sync, and thus fixes the cancellation issue.

joshheald commented:
@koke c0b6e12 fixes the cancellation issue. It was happening due to `cancelPayment` being called before `collectPayment` action was dispatched, while we were waiting for an unnecessary `syncPlugins` action. 

Removing the extra sync fixed the cancelation, and makes payment preparation a little quicker.
